File: app/src/main/java/network/ex03.java
package network;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class ex03 {
    private static final int PORT = 8888;
    private static final List<PrintWriter> CLIENTS = Collections.synchronizedList(new ArrayList<>());

    public static void main(String[] args) {
        System.out.println("chat server template (port " + PORT + ")");
        System.out.println("use startServer() to run a simple multi-client chat server");
    }

    static void startServer() {
        ExecutorService pool = Executors.newFixedThreadPool(10);

        try (ServerSocket serverSocket = new ServerSocket(PORT)) {
            while (!Thread.currentThread().isInterrupted()) {
                Socket socket = serverSocket.accept();
                pool.execute(new ClientHandler(socket));
            }
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            pool.shutdown();
        }
    }

    static void broadcast(String message) {
        synchronized (CLIENTS) {
            for (PrintWriter writer : CLIENTS) {
                writer.println(message);
            }
        }
    }

    static class ClientHandler implements Runnable {
        private final Socket socket;
        private PrintWriter out;
        private String name;

        ClientHandler(Socket socket) {
            this.socket = socket;
        }

        @Override
        public void run() {
            try (BufferedReader in = new BufferedReader(new InputStreamReader(socket.getInputStream()))) {
                out = new PrintWriter(socket.getOutputStream(), true);
                name = in.readLine();
                CLIENTS.add(out);
                broadcast("[" + name + "] joined");

                String message;
                while ((message = in.readLine()) != null) {
                    broadcast("[" + name + "]: " + message);
                }
            } catch (IOException e) {
                System.out.println("connection closed: " + name);
            } finally {
                if (out != null) {
                    CLIENTS.remove(out);
                }
                broadcast("[" + name + "] left");
                try {
                    socket.close();
                } catch (IOException ignored) {
                }
            }
        }
    }
}

File: app-kotlin/src/main/kotlin/network/ex03.kt
package network

import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.io.PrintWriter
import java.net.ServerSocket
import java.net.Socket
import java.util.Collections
import java.util.concurrent.Executors

class ex03 {
    companion object {
        private const val PORT = 8888
        private val clients = Collections.synchronizedList(mutableListOf<PrintWriter>())

        @JvmStatic
        fun main(args: Array<String>) {
            println("chat server template (port $PORT)")
            println("use startServer() to run a simple multi-client chat server")
        }

        fun startServer() {
            val pool = Executors.newFixedThreadPool(10)
            try {
                ServerSocket(PORT).use { serverSocket ->
                    while (!Thread.currentThread().isInterrupted) {
                        val socket = serverSocket.accept()
                        pool.execute(ClientHandler(socket))
                    }
                }
            } catch (e: IOException) {
                e.printStackTrace()
            } finally {
                pool.shutdown()
            }
        }

        private fun broadcast(message: String) {
            synchronized(clients) {
                clients.forEach { it.println(message) }
            }
        }

        private class ClientHandler(private val socket: Socket) : Runnable {
            private var out: PrintWriter? = null
            private var name: String = "anonymous"

            override fun run() {
                try {
                    BufferedReader(InputStreamReader(socket.getInputStream())).use { reader ->
                        out = PrintWriter(socket.getOutputStream(), true)
                        name = reader.readLine() ?: "anonymous"
                        clients.add(out)
                        broadcast("[$name] joined")

                        while (true) {
                            val message = reader.readLine() ?: break
                            broadcast("[$name]: $message")
                        }
                    }
                } catch (_: IOException) {
                    println("connection closed: $name")
                } finally {
                    out?.let { clients.remove(it) }
                    broadcast("[$name] left")
                    try {
                        socket.close()
                    } catch (_: IOException) {
                    }
                }
            }
        }
    }
}
